server/datastore/mysql/software_installers.go

```diff
@@ -0,0 +1,139 @@
+package mysql
+
+import (
+	"context"
+
+	"github.com/fleetdm/fleet/v4/server/contexts/ctxerr"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/jmoiron/sqlx"
+)
+
+func (ds *Datastore) MatchOrCreateSoftwareInstaller(ctx context.Context, payload *fleet.UploadSoftwareInstallerPayload) (uint, error) {
+	// TODO: confirm ds context is appropriate in each of the following (should we use writer for all?)
```

Comment:
I think this is correct, those are all different/independent entities so you don't need to read a write immediately after. Only case would be of install script is the same as post-install script (it would try to create the script twice and fail), but that's not realistic.